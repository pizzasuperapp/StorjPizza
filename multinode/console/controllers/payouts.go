// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

package controllers

import (
	"encoding/json"
	"net/http"

	"github.com/gorilla/mux"
	"github.com/zeebo/errs"
	"go.uber.org/zap"

	"storj.io/common/storj"
	"storj.io/storj/multinode/nodes"
	"storj.io/storj/multinode/payouts"
)

var (
	// ErrPayouts is an internal error type for payouts web api controller.
	ErrPayouts = errs.Class("payouts web api controller")
)

// Payouts is a web api controller.
type Payouts struct {
	log     *zap.Logger
	service *payouts.Service
}

// NewPayouts is a constructor for Payouts.
func NewPayouts(log *zap.Logger, service *payouts.Service) *Payouts {
	return &Payouts{
		log:     log,
		service: service,
	}
}

// Earned handles retrieval total earned amount .
func (controller *Payouts) Earned(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	earned, err := controller.service.Earned(ctx)
	if err != nil {
		controller.log.Error("all node total earned internal error", zap.Error(err))
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(earned); err != nil {
		controller.log.Error("failed to write json response", zap.Error(err))
		return
	}
}

// NodeExpectations handles node's estimated and undistributed.
func (controller *Payouts) NodeExpectations(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	w.Header().Add("Content-Type", "application/json")
	segmentParams := mux.Vars(r)

	id, ok := segmentParams["nodeID"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable nodeID"))
		return
	}

	nodeID, err := storj.NodeIDFromString(id)
	if err != nil {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.Wrap(err))
		return
	}

	expectations, err := controller.service.NodeExpectations(ctx, nodeID)
	if err != nil {
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(expectations); err != nil {
		controller.log.Error("failed to write json response", zap.Error(err))
		return
	}
}

// Expectations handles nodes estimated and undistributed earnings.
func (controller *Payouts) Expectations(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	w.Header().Add("Content-Type", "application/json")

	expectations, err := controller.service.Expectations(ctx)
	if err != nil {
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(expectations); err != nil {
		controller.log.Error("failed to write json response", zap.Error(err))
		return
	}
}

// SummaryPeriod handles retrieval from nodes for specific period.
func (controller *Payouts) SummaryPeriod(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	w.Header().Add("Content-Type", "application/json")
	segmentParams := mux.Vars(r)

	period, ok := segmentParams["period"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable period"))
		return
	}

	summary, err := controller.service.SummaryPeriod(ctx, period)
	if err != nil {
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(summary); err != nil {
		controller.log.Error("failed to write json response", zap.Error(err))
		return
	}
}

// Summary handles retrieval from nodes.
func (controller *Payouts) Summary(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	w.Header().Add("Content-Type", "application/json")

	summary, err := controller.service.Summary(ctx)
	if err != nil {
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(summary); err != nil {
		controller.log.Error("failed to write json response", zap.Error(err))
		return
	}
}

// SummarySatellitePeriod handles retrieval from nodes from specific satellite for specific period.
func (controller *Payouts) SummarySatellitePeriod(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	w.Header().Add("Content-Type", "application/json")
	segmentParams := mux.Vars(r)

	period, ok := segmentParams["period"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable period"))
		return
	}

	id, ok := segmentParams["id"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable satellite id"))
		return
	}

	satelliteID, err := storj.NodeIDFromString(id)
	if err != nil {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.Wrap(err))
		return
	}

	summary, err := controller.service.SummarySatellitePeriod(ctx, satelliteID, period)
	if err != nil {
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(summary); err != nil {
		controller.log.Error("failed to write json response", zap.Error(err))
		return
	}
}

// SummarySatellite handles retrieval from nodes from specific satellite.
func (controller *Payouts) SummarySatellite(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	w.Header().Add("Content-Type", "application/json")
	segmentParams := mux.Vars(r)

	id, ok := segmentParams["id"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable satellite id"))
		return
	}

	satelliteID, err := storj.NodeIDFromString(id)
	if err != nil {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.Wrap(err))
		return
	}

	summary, err := controller.service.SummarySatellite(ctx, satelliteID)
	if err != nil {
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(summary); err != nil {
		controller.log.Error("failed to write json response", zap.Error(err))
		return
	}
}

// Paystub returns all summed paystubs.
func (controller *Payouts) Paystub(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	w.Header().Add("Content-Type", "application/json")
	segmentParams := mux.Vars(r)

	nodeIDstring, ok := segmentParams["nodeID"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable nodeID"))
		return
	}

	nodeID, err := storj.NodeIDFromString(nodeIDstring)
	if err != nil {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.Wrap(err))
		return
	}

	paystub, err := controller.service.Paystub(ctx, nodeID)
	if err != nil {
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(paystub); err != nil {
		controller.log.Error("failed to write json response", zap.Error(err))
		return
	}
}

// PaystubSatellite returns all summed paystubs from specific satellite.
func (controller *Payouts) PaystubSatellite(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	w.Header().Add("Content-Type", "application/json")
	segmentParams := mux.Vars(r)

	id, ok := segmentParams["id"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable satellite id"))
		return
	}

	satelliteID, err := storj.NodeIDFromString(id)
	if err != nil {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.Wrap(err))
		return
	}

	nodeIDstring, ok := segmentParams["nodeID"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable nodeID"))
		return
	}

	nodeID, err := storj.NodeIDFromString(nodeIDstring)
	if err != nil {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.Wrap(err))
		return
	}

	paystub, err := controller.service.PaystubSatellite(ctx, nodeID, satelliteID)
	if err != nil {
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(paystub); err != nil {
		controller.log.Error("failed to write json response", zap.Error(err))
		return
	}
}

// PaystubSatellitePeriod returns satellite summed paystubs for period.
func (controller *Payouts) PaystubSatellitePeriod(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	w.Header().Add("Content-Type", "application/json")
	segmentParams := mux.Vars(r)

	id, ok := segmentParams["id"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable satellite id"))
		return
	}

	satelliteID, err := storj.NodeIDFromString(id)
	if err != nil {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.Wrap(err))
		return
	}

	nodeIDstring, ok := segmentParams["nodeID"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable nodeID"))
		return
	}

	nodeID, err := storj.NodeIDFromString(nodeIDstring)
	if err != nil {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.Wrap(err))
		return
	}

	period, ok := segmentParams["period"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable period"))
		return
	}

	paystub, err := controller.service.PaystubSatellitePeriod(ctx, period, nodeID, satelliteID)
	if err != nil {
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(paystub); err != nil {
		controller.log.Error("failed to write json response", zap.Error(err))
		return
	}
}

// PaystubPeriod returns all satellites summed paystubs for period.
func (controller *Payouts) PaystubPeriod(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	w.Header().Add("Content-Type", "application/json")
	segmentParams := mux.Vars(r)

	nodeIDstring, ok := segmentParams["nodeID"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable nodeID"))
		return
	}

	nodeID, err := storj.NodeIDFromString(nodeIDstring)
	if err != nil {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.Wrap(err))
		return
	}

	period, ok := segmentParams["period"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("couldn't receive route variable period"))
		return
	}

	paystub, err := controller.service.PaystubPeriod(ctx, period, nodeID)
	if err != nil {
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(paystub); err != nil {
		controller.log.Error("failed to write json response", zap.Error(err))
		return
	}
}

// HeldAmountSummary handles retrieving held amount history for a node.
func (controller *Payouts) HeldAmountSummary(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	defer mon.Task()(&ctx)(&err)

	vars := mux.Vars(r)

	idString, ok := vars["nodeID"]
	if !ok {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.New("node id segment parameter is missing"))
		return
	}
	nodeID, err := storj.NodeIDFromString(idString)
	if err != nil {
		controller.serveError(w, http.StatusBadRequest, ErrPayouts.Wrap(err))
		return
	}

	heldSummary, err := controller.service.HeldAmountSummary(ctx, nodeID)
	if err != nil {
		if nodes.ErrNoNode.Has(err) {
			controller.serveError(w, http.StatusNotFound, ErrPayouts.Wrap(err))
			return
		}

		controller.log.Error("held amount history internal error", zap.Error(ErrPayouts.Wrap(err)))
		controller.serveError(w, http.StatusInternalServerError, ErrPayouts.Wrap(err))
		return
	}

	if err = json.NewEncoder(w).Encode(heldSummary); err != nil {
		controller.log.Error("failed to write json response", zap.Error(ErrPayouts.Wrap(err)))
		return
	}
}

// serveError set http statuses and send json error.
func (controller *Payouts) serveError(w http.ResponseWriter, status int, err error) {
	w.WriteHeader(status)

	var response struct {
		Error string `json:"error"`
	}

	response.Error = err.Error()

	err = json.NewEncoder(w).Encode(response)
	if err != nil {
		controller.log.Error("failed to write json error response", zap.Error(err))
	}
}
