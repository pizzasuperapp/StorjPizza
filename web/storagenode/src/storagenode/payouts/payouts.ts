// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import { getMonthsBeforeNow } from '@/app/utils/payout';

/**
 * Exposes all payout-related functionality.
 */
export interface PayoutApi {
    /**
     * Fetches paystubs by selected period.
     * @throws Error
     */
    getPaystubsForPeriod(paymentInfoParameters: PaymentInfoParameters): Promise<Paystub[]>;

    /**
     * Fetches available payout periods.
     * @throws Error
     */
    getPayoutPeriods(id: string): Promise<PayoutPeriod[]>;

    /**
     * Fetches held history for all satellites.
     * @throws Error
     */
    getHeldHistory(): Promise<SatelliteHeldHistory[]>;

    /**
     * Fetch estimated payout information.
     * @throws Error
     */
    getEstimatedPayout(satelliteId: string): Promise<EstimatedPayout>;

    /**
     * Fetches payout history for all satellites.
     * @throws Error
     */
    getPayoutHistory(period: string): Promise<SatellitePayoutForPeriod[]>;
}

// TODO: move to config.
/**
 * Divider to convert payout amounts to cents.
 */
const PRICE_DIVIDER = 10000;

/**
 * Represents payout period month and year.
 */
export class PayoutPeriod {
    public constructor(
        public year: number = new Date().getUTCFullYear(),
        public month: number = new Date().getUTCMonth(),
    ) {}

    public get period(): string {
        return this.month < 9 ? `${this.year}-0${this.month + 1}` : `${this.year}-${this.month + 1}`;
    }

    /**
     * Parses PayoutPeriod from string.
     * @param period string
     */
    public static fromString(period: string): PayoutPeriod {
        const periodArray = period.split('-');

        return new PayoutPeriod(parseInt(periodArray[0]), parseInt(periodArray[1]) - 1);
    }
}

/**
 * Holds request arguments for payout information.
 */
export class PaymentInfoParameters {
    public constructor(
        public start: PayoutPeriod | null = null,
        public end: PayoutPeriod = new PayoutPeriod(),
        public satelliteId: string = '',
    ) {}
}

/**
 * PayStub is an entity that holds usage and cash amounts that will be paid to storagenode operator after for month period.
 */
export class Paystub {
    public constructor(
        public usageAtRest: number = 0,
        public usageGet: number = 0,
        public usagePut: number = 0,
        public usageGetRepair: number = 0,
        public usagePutRepair: number = 0,
        public usageGetAudit: number = 0,
        public compAtRest: number = 0,
        public compGet: number = 0,
        public compPut: number = 0,
        public compGetRepair: number = 0,
        public compPutRepair: number = 0,
        public compGetAudit: number = 0,
        public surgePercent: number = 0,
        public held: number = 0,
        public owed: number = 0,
        public disposed: number = 0,
        public paid: number = 0,
        public distributed: number = 0,
    ) {}

    /**
     * Returns payout amount multiplier.
     */
    public get surgeMultiplier(): number {
        // 0 in backend uses instead of "without multiplier"
        return this.surgePercent === 0 ? 1 : this.surgePercent / 100;
    }
}

/**
 * Summary of paystubs by period.
 * Payout amounts converted to cents.
 */
export class TotalPaystubForPeriod {
    public usageAtRest = 0;
    public usageGet = 0;
    public usagePut = 0;
    public usageGetRepair = 0;
    public usagePutRepair = 0;
    public usageGetAudit = 0;
    public compAtRest = 0;
    public compGet = 0;
    public compPut = 0;
    public compGetRepair = 0;
    public compPutRepair = 0;
    public compGetAudit = 0;
    public surgePercent = 0;
    public held = 0;
    public owed = 0;
    public disposed = 0;
    public paid = 0;
    public paidWithoutSurge = 0;
    public grossWithSurge = 0;
    public distributed = 0;

    public constructor(
        paystubs: Paystub[] = [],
    ) {
        paystubs.forEach(paystub => {
            this.usageAtRest += paystub.usageAtRest;
            this.usageGet += paystub.usageGet;
            this.usagePut += paystub.usagePut;
            this.usageGetRepair += paystub.usageGetRepair;
            this.usagePutRepair += paystub.usagePutRepair;
            this.usageGetAudit += paystub.usageGetAudit;
            this.compAtRest += this.convertToCents(paystub.compAtRest);
            this.compGet += this.convertToCents(paystub.compGet);
            this.compPut += this.convertToCents(paystub.compPut);
            this.compGetRepair += this.convertToCents(paystub.compGetRepair);
            this.compPutRepair += this.convertToCents(paystub.compPutRepair);
            this.compGetAudit += this.convertToCents(paystub.compGetAudit);
            this.held += this.convertToCents(paystub.held);
            this.owed += this.convertToCents(paystub.owed);
            this.disposed += this.convertToCents(paystub.disposed);
            this.paid += this.convertToCents(paystub.paid);
            this.surgePercent = paystub.surgePercent;
            this.paidWithoutSurge += this.convertToCents(paystub.paid + paystub.held - paystub.disposed) / paystub.surgeMultiplier;
            this.grossWithSurge += this.convertToCents(paystub.paid + paystub.held - paystub.disposed);
            this.distributed += this.convertToCents(paystub.distributed);
        });
    }

    private convertToCents(value: number): number {
        return value / PRICE_DIVIDER;
    }
}

/**
 * Holds accumulated held and earned payouts.
 */
export class TotalPayments {
    public held = 0;
    public paid = 0;
    public disposed = 0;
    // TODO: remove
    public currentMonthEarnings = 0;
    public balance = 0;

    public constructor(
        paystubs: Paystub[] = [],
    ) {
        paystubs.forEach(paystub => {
            this.paid += this.convertToCents(paystub.paid);
            this.disposed += this.convertToCents(paystub.disposed);
            this.held += this.convertToCents(paystub.held - paystub.disposed);
            this.balance += this.convertToCents(paystub.paid - paystub.distributed);
        });
    }

    public setCurrentMonthEarnings(value: number): void {
        this.currentMonthEarnings = value;
    }

    private convertToCents(value: number): number {
        return value / PRICE_DIVIDER;
    }
}

/**
 * Holds held history information for all satellites.
 */
export class SatelliteHeldHistory {
    public constructor(
        public satelliteID: string = '',
        public satelliteName: string = '',
        public holdForFirstPeriod: number = 0,
        public holdForSecondPeriod: number = 0,
        public holdForThirdPeriod: number = 0,
        public totalHeld: number = 0,
        public totalDisposed: number = 0,
        public joinedAt: Date = new Date(),
    ) {
        this.totalHeld = this.convertToCents(this.totalHeld - this.totalDisposed);
        this.totalDisposed = this.convertToCents(this.totalDisposed);
        this.holdForFirstPeriod = this.convertToCents(this.holdForFirstPeriod);
        this.holdForSecondPeriod = this.convertToCents(this.holdForSecondPeriod);
        this.holdForThirdPeriod = this.convertToCents(this.holdForThirdPeriod);
    }

    public get monthsWithNode(): number {
        return getMonthsBeforeNow(this.joinedAt);
    }

    private convertToCents(value: number): number {
        return value / PRICE_DIVIDER;
    }
}

/**
 * Contains estimated payout information for current and last periods.
 */
export class EstimatedPayout {
    public constructor(
        public currentMonth: PreviousMonthEstimatedPayout = new PreviousMonthEstimatedPayout(),
        public previousMonth: PreviousMonthEstimatedPayout = new PreviousMonthEstimatedPayout(),
        public currentMonthExpectations: number = 0,
    ) {}
}

/**
 * Contains last month estimated payout information.
 */
export class PreviousMonthEstimatedPayout {
    public constructor(
        public egressBandwidth: number = 0,
        public egressBandwidthPayout: number = 0,
        public egressRepairAudit: number = 0,
        public egressRepairAuditPayout: number = 0,
        public diskSpace: number = 0,
        public diskSpacePayout: number = 0,
        public heldRate: number = 0,
        public payout: number = 0,
        public held: number = 0,
    ) {}
}

/**
 * Contains payout information for payout history table.
 */
export class SatellitePayoutForPeriod {
    public constructor(
        public satelliteID: string = '',
        public satelliteName: string = '',
        public age: number = 1,
        public earned: number = 0,
        public surge: number = 0,
        public surgePercent: number = 0,
        public held: number = 0,
        public afterHeld: number = 0,
        public disposed: number = 0,
        public paid: number = 0,
        public receipt: string = '',
        public isExitComplete: boolean = false,
        public heldPercent: number = 0,
        public distributed: number = 0,
    ) {
        this.earned = this.convertToCents(this.earned);
        this.surge = this.convertToCents(this.surge);
        this.held = this.convertToCents(this.held);
        this.afterHeld = this.convertToCents(this.afterHeld);
        this.disposed = this.convertToCents(this.disposed);
        this.paid = this.convertToCents(this.paid);
        this.distributed = this.convertToCents(this.distributed);
    }

    public get transactionLink(): string {
        const prefixed = function (hash: string): string {
            if (hash.indexOf("0x") != 0) {
                return "0x" + hash;
            }
            return hash;
        }
        if (!this.receipt) {
            return '';
        }

        if (this.receipt.indexOf('eth') !== -1) {
            return `https://etherscan.io/tx/${prefixed(this.receipt.slice(4))}`;
        }

        {
            const zkScanUrl = 'https://zkscan.io/explorer/transactions'

            if (this.receipt.indexOf('zksync') !== -1) {
                return `${zkScanUrl}/${prefixed(this.receipt.slice(7))}`;
            }
            if (this.receipt.indexOf('zkwithdraw') !== -1) {
                return `${zkScanUrl}/${prefixed(this.receipt.slice(11))}`;
            }
        }

        if (this.receipt.indexOf('polygon') !== -1) {
            return `https://polygonscan.com/tx/${prefixed(this.receipt.slice(8))}`;
        }

        return this.receipt;
    }

    private convertToCents(value: number): number {
        return value / PRICE_DIVIDER;
    }
}
