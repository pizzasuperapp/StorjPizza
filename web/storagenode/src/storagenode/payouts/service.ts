// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    EstimatedPayout,
    PaymentInfoParameters,
    PayoutApi,
    PayoutPeriod,
    Paystub,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod,
    TotalPayments,
    TotalPaystubForPeriod,
} from '@/storagenode/payouts/payouts';

/**
 * PayoutService is used to store and handle node paystub information.
 * PayoutService exposes a business logic related to payouts.
 */
export class PayoutService {
    private readonly payouts: PayoutApi;

    public constructor(api: PayoutApi) {
        this.payouts = api;
    }

    /**
     * Gets summary of paystubs for given period.
     * @param start period start
     * @param end period end
     * @param satelliteId
     */
    public async paystubSummaryForPeriod(start: PayoutPeriod | null, end: PayoutPeriod, satelliteId?: string): Promise<TotalPaystubForPeriod> {
        const paystubs: Paystub[] = await this.payouts.getPaystubsForPeriod(new PaymentInfoParameters(start, end, satelliteId));

        return new TotalPaystubForPeriod(paystubs);
    }

    /**
     * Gets held and paid summary for given period.
     * @param start period start
     * @param end period end
     * @param satelliteId
     */
    public async totalPayments(start: PayoutPeriod, end: PayoutPeriod, satelliteId: string): Promise<TotalPayments> {
        const paystubs: Paystub[] = await this.payouts.getPaystubsForPeriod(new PaymentInfoParameters(start, end, satelliteId));

        return new TotalPayments(paystubs);
    }

    /**
     * Gets list of payout periods that have paystubs for selected satellite.
     * If satelliteId is not provided returns periods for all satellites.
     * @param satelliteId
     */
    public async availablePeriods(satelliteId: string): Promise<PayoutPeriod[]> {
        return await this.payouts.getPayoutPeriods(satelliteId);
    }

    /**
     * Gets list of payout history items for given period by satellites.
     * @param payoutHistoryPeriod year and month representation
     */
    public async payoutHistory(payoutHistoryPeriod: string): Promise<SatellitePayoutForPeriod[]> {
        return await this.payouts.getPayoutHistory(payoutHistoryPeriod);
    }

    /**
     * Gets list of held history for all satellites.
     */
    public async allSatellitesHeldHistory(): Promise<SatelliteHeldHistory[]> {
        return await this.payouts.getHeldHistory();
    }

    /**
     * Gets estimated payout when no data in paystub.
     * @param satelliteId
     */
    public async estimatedPayout(satelliteId: string): Promise<EstimatedPayout> {
        return await this.payouts.getEstimatedPayout(satelliteId);
    }
}
