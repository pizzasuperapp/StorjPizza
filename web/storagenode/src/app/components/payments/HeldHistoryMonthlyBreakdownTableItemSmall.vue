// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="held-history-table-container--small__item">
        <div class="held-history-table-container--small__item__satellite-info">
            <div>
                <p class="held-history-table-container--small__item__satellite-info__name">{{ heldHistoryItem.satelliteName }}</p>
                <p class="held-history-table-container--small__item__satellite-info__months">{{ heldHistoryItem.monthsWithNode }} month</p>
            </div>
            <div class="held-history-table-container--small__item__satellite-info__button">
                <div v-if="isExpanded" class="icon hide" @click="hide">
                    <blue-hide-icon />
                </div>
                <div v-else class="icon expand" @click="expand">
                    <blue-expand-icon />
                </div>
            </div>
        </div>
        <transition name="fade">
            <div v-if="isExpanded" class="held-history-table-container--small__item__held-info">
                <div class="held-history-table-container--small__item__held-info__item">
                    <p class="held-history-table-container--small__item__held-info__item__label">Month 1-3</p>
                    <p class="held-history-table-container--small__item__held-info__item__value">{{ heldHistoryItem.holdForFirstPeriod | centsToDollars }}</p>
                </div>
                <div class="held-history-table-container--small__item__held-info__item">
                    <p class="held-history-table-container--small__item__held-info__item__label">Month 4-6</p>
                    <p class="held-history-table-container--small__item__held-info__item__value">{{ heldHistoryItem.holdForSecondPeriod | centsToDollars }}</p>
                </div>
                <div class="held-history-table-container--small__item__held-info__item">
                    <p class="held-history-table-container--small__item__held-info__item__label">Month 7-9</p>
                    <p class="held-history-table-container--small__item__held-info__item__value">{{ heldHistoryItem.holdForThirdPeriod | centsToDollars }}</p>
                </div>
            </div>
        </transition>
    </div>
</template>

<script lang="ts">
import { Component, Prop } from 'vue-property-decorator';

import BlueHideIcon from '@/../static/images/common/BlueMinus.svg';
import BlueExpandIcon from '@/../static/images/common/BluePlus.svg';

import BaseSmallHeldHistoryTable from '@/app/components/payments/BaseSmallHeldHistoryTable.vue';

import { SatelliteHeldHistory } from '@/storagenode/payouts/payouts';

// @vue/component
@Component({
    components: {
        BlueHideIcon,
        BlueExpandIcon,
    },
})
export default class HeldHistoryMonthlyBreakdownTableSmall extends BaseSmallHeldHistoryTable {
    @Prop({default: () => new SatelliteHeldHistory()})
    public readonly heldHistoryItem: SatelliteHeldHistory;
}
</script>
