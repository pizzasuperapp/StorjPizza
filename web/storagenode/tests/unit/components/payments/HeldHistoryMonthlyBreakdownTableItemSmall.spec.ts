// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import HeldHistoryMonthlyBreakdownTableItemSmall from '@/app/components/payments/HeldHistoryMonthlyBreakdownTableItemSmall.vue';

import { SatelliteHeldHistory } from '@/storagenode/payouts/payouts';
import { createLocalVue, shallowMount } from '@vue/test-utils';

const localVue = createLocalVue();

localVue.filter('centsToDollars', (cents: number): string => {
    return `$${(cents / 100).toFixed(2)}`;
});

describe('HeldHistoryMonthlyBreakdownTableSmall', (): void => {
    it('renders correctly with actual values',  async (): Promise<void> => {
        const _Date = Date;
        const testJoinAt = new Date(Date.UTC(2019, 6, 30));
        const mockedDate = new Date(1580522290000);
        global.Date = jest.fn(() => mockedDate);

        const wrapper = shallowMount(HeldHistoryMonthlyBreakdownTableItemSmall, {
            propsData: {
                heldHistoryItem: new SatelliteHeldHistory(
                    '1',
                    'name1',
                    6,
                    50000,
                    7333880,
                    7852235,
                    757576,
                    testJoinAt,
                ),
            },
            localVue,
        });

        expect(wrapper).toMatchSnapshot();

        wrapper.find('.expand').trigger('click');

        await localVue.nextTick();

        expect(wrapper).toMatchSnapshot();

        wrapper.find('.hide').trigger('click');

        await localVue.nextTick();

        expect(wrapper).toMatchSnapshot();

        global.Date = _Date;
    });
});
