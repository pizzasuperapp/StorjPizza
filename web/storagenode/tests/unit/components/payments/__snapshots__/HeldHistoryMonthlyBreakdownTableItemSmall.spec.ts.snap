// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeldHistoryMonthlyBreakdownTableSmall renders correctly with actual values 1`] = `
<div class="held-history-table-container--small__item">
  <div class="held-history-table-container--small__item__satellite-info">
    <div>
      <p class="held-history-table-container--small__item__satellite-info__name">name1</p>
      <p class="held-history-table-container--small__item__satellite-info__months">8 month</p>
    </div>
    <div class="held-history-table-container--small__item__satellite-info__button">
      <div class="icon expand">
        <blue-expand-icon-stub></blue-expand-icon-stub>
      </div>
    </div>
  </div>
  <transition-stub name="fade">
    <!---->
  </transition-stub>
</div>
`;

exports[`HeldHistoryMonthlyBreakdownTableSmall renders correctly with actual values 2`] = `
<div class="held-history-table-container--small__item">
  <div class="held-history-table-container--small__item__satellite-info">
    <div>
      <p class="held-history-table-container--small__item__satellite-info__name">name1</p>
      <p class="held-history-table-container--small__item__satellite-info__months">8 month</p>
    </div>
    <div class="held-history-table-container--small__item__satellite-info__button">
      <div class="icon hide">
        <blue-hide-icon-stub></blue-hide-icon-stub>
      </div>
    </div>
  </div>
  <transition-stub name="fade">
    <div class="held-history-table-container--small__item__held-info">
      <div class="held-history-table-container--small__item__held-info__item">
        <p class="held-history-table-container--small__item__held-info__item__label">Month 1-3</p>
        <p class="held-history-table-container--small__item__held-info__item__value">$0.00</p>
      </div>
      <div class="held-history-table-container--small__item__held-info__item">
        <p class="held-history-table-container--small__item__held-info__item__label">Month 4-6</p>
        <p class="held-history-table-container--small__item__held-info__item__value">$0.05</p>
      </div>
      <div class="held-history-table-container--small__item__held-info__item">
        <p class="held-history-table-container--small__item__held-info__item__label">Month 7-9</p>
        <p class="held-history-table-container--small__item__held-info__item__value">$7.33</p>
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`HeldHistoryMonthlyBreakdownTableSmall renders correctly with actual values 3`] = `
<div class="held-history-table-container--small__item">
  <div class="held-history-table-container--small__item__satellite-info">
    <div>
      <p class="held-history-table-container--small__item__satellite-info__name">name1</p>
      <p class="held-history-table-container--small__item__satellite-info__months">8 month</p>
    </div>
    <div class="held-history-table-container--small__item__satellite-info__button">
      <div class="icon expand">
        <blue-expand-icon-stub></blue-expand-icon-stub>
      </div>
    </div>
  </div>
  <transition-stub name="fade">
    <!---->
  </transition-stub>
</div>
`;
